File: app/src/main/java/com/example/thekhaeng/viewstatesavetest/view/java/base/BaseViewGroup.java
package com.example.thekhaeng.viewstatesavetest.view.java.base;

import android.content.Context;
import android.os.Build;
import android.os.Parcel;
import android.os.Parcelable;
import android.support.annotation.RequiresApi;
import android.util.AttributeSet;
import android.util.SparseArray;
import android.widget.FrameLayout;

/**
 * Created by TheKhaeng on 9/22/2016.
 */

@SuppressWarnings( "unchecked" )
public abstract class BaseViewGroup extends FrameLayout{


    public BaseViewGroup( Context context ){
        super( context );
        setup( null, 0, 0 );
    }

    public BaseViewGroup( Context context, AttributeSet attrs ){
        super( context, attrs );
        setup( attrs, 0, 0 );
    }

    public BaseViewGroup( Context context, AttributeSet attrs, int defStyleAttr ){
        super( context, attrs, defStyleAttr );
        setup( attrs, defStyleAttr, 0 );
    }

    @RequiresApi( api = Build.VERSION_CODES.LOLLIPOP )
    public BaseViewGroup( Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes ){
        super( context, attrs, defStyleAttr, defStyleRes );
        setup( attrs, defStyleAttr, defStyleRes );
    }

    private void setup( AttributeSet attrs, int defStyleAttr, int defStyleRes ){
        if( attrs != null ) setupStyleables( attrs, defStyleAttr, defStyleRes );
        inflateLayout();
        bindView();
        setupInstance();
        setupView();
    }

    /**
     * Custom handle SavedState child to "fix restore state in same resource id"
     * when use ViewGroup more than one.
     * <p>
     * the method must be call in subclass.
     **/
    protected Parcelable onSaveChildInstanceState( ChildSavedState ss ){
        ss.childrenStates = new SparseArray();
        for( int i = 0; i < getChildCount(); i++ ){
            int id = getChildAt( i ).getId();
            if( id != 0 ){
                SparseArray childrenState = new SparseArray();
                getChildAt( i ).saveHierarchyState( childrenState );
                ss.childrenStates.put( id, childrenState );
            }
        }
        return ss;
    }

    @Override
    protected void onRestoreInstanceState( Parcelable state ){
        if( !( state instanceof ChildSavedState ) ){
            super.onRestoreInstanceState( state );
            return;
        }
        ChildSavedState ss = (ChildSavedState) state;
        super.onRestoreInstanceState( ss );
        onRestoreChildInstanceState( ss );
    }

    private void onRestoreChildInstanceState( ChildSavedState ss ){
        for( int i = 0; i < getChildCount(); i++ ){
            int id = getChildAt( i ).getId();
            if( id != 0 ){
                if( ss.childrenStates.get( id ) != null ){
                    SparseArray childrenState = (SparseArray) ss.childrenStates.get( id );
                    getChildAt( i ).restoreHierarchyState( childrenState );
                }
            }
        }
    }


    public abstract static class ChildSavedState extends BaseSavedState{
        SparseArray childrenStates;

        public ChildSavedState( Parcelable superState ){
            super( superState );
        }

        public ChildSavedState( Parcel in, ClassLoader classLoader ){
            super( in );
            childrenStates = in.readSparseArray( classLoader );
        }

        @Override
        public void writeToParcel( Parcel out, int flags ){
            super.writeToParcel( out, flags );
            out.writeSparseArray( childrenStates );
        }
    }


    @Override
    protected void dispatchSaveInstanceState( SparseArray<Parcelable> container ){
        dispatchFreezeSelfOnly( container );
    }

    @Override
    protected void dispatchRestoreInstanceState( SparseArray<Parcelable> container ){
        dispatchThawSelfOnly( container );
    }

    private void inflateLayout(){
        inflate( getContext(), getLayoutRes(), this );
    }

    protected abstract int getLayoutRes();

    protected void setupStyleables( AttributeSet attrs, int defStyleAttr, int defStyleRes ){
    }

    protected void bindView(){
    }

    protected void setupInstance(){
    }

    protected abstract void setupView();

}

File: app/src/main/java/com/example/thekhaeng/viewstatesavetest/view/kotlin/base/BaseViewGroup.kt
package com.example.thekhaeng.viewstatesavetest.view.kotlin.base

import android.content.Context
import android.os.Parcel
import android.os.Parcelable
import android.util.AttributeSet
import android.util.SparseArray
import android.view.View
import android.widget.FrameLayout

/**
 * Created by TheKhaeng on 9/22/2016.
 */

abstract class BaseViewGroup
@JvmOverloads constructor(context: Context,
                          attrs: AttributeSet? = null,
                          defStyleAttr: Int = 0,
                          defStyleRes: Int = 0)
    : FrameLayout(context, attrs, defStyleAttr, defStyleRes) {

    init {
        setup(attrs, defStyleAttr, defStyleRes)
    }

    private fun setup(attrs: AttributeSet?, defStyleAttr: Int, defStyleRes: Int) {
        if (attrs != null) setupStyleables(attrs, defStyleAttr, defStyleRes)
        inflateLayout()
        bindView()
        setupInstance()
        setupView()
    }

    /**
     * Custom handle SavedState child to "fix restore state in same resource id"
     * when use ViewGroup more than one.
     *
     *
     * the method must be call in subclass.
     */
    protected fun onSaveChildInstanceState(ss: ChildSavedState): Parcelable {
        ss.childrenStates = SparseArray()
        for (i in 0 until childCount) {
            val id = getChildAt(i).id
            if (id != 0) {
                val childrenState = SparseArray<Parcelable>()
                getChildAt(i).saveHierarchyState(childrenState)
                ss.childrenStates?.put(id, childrenState)
            }

        }
        return ss
    }

    override fun onRestoreInstanceState(state: Parcelable) {
        if (state !is ChildSavedState) {
            super.onRestoreInstanceState(state)
            return
        }
        val ss = state
        super.onRestoreInstanceState(ss.superState)
        onRestoreChildInstanceState(ss)
    }

    private fun onRestoreChildInstanceState(ss: ChildSavedState) {
        for (i in 0 until childCount) {
            val id = getChildAt(i).id
            if (id != 0) {
                if (ss.childrenStates?.get(id) != null) {
                    val childrenState = ss.childrenStates?.get(id) as SparseArray<Parcelable?>
                    getChildAt(i).restoreHierarchyState(childrenState)
                }
            }
        }
    }


    abstract class ChildSavedState : View.BaseSavedState {
        internal var childrenStates: SparseArray<Any>? = null

        constructor(superState: Parcelable) : super(superState) {}

        constructor(`in`: Parcel, classLoader: ClassLoader) : super(`in`) {
            childrenStates = `in`.readSparseArray(classLoader)
        }

        override fun writeToParcel(out: Parcel, flags: Int) {
            super.writeToParcel(out, flags)
            out.writeSparseArray(childrenStates)
        }
    }


    override fun dispatchSaveInstanceState(container: SparseArray<Parcelable>) {
        dispatchFreezeSelfOnly(container)
    }

    override fun dispatchRestoreInstanceState(container: SparseArray<Parcelable>) {
        dispatchThawSelfOnly(container)
    }

    private fun inflateLayout() {
        View.inflate(context, getLayoutRes(), this)
    }

    protected abstract fun getLayoutRes(): Int

    protected abstract fun setupView()

    protected open fun setupStyleables(attrs: AttributeSet, defStyleAttr: Int, defStyleRes: Int) {}

    protected open fun bindView() {}

    protected fun setupInstance() {}

}
